File: java/core/src/main/java/com/github/fernthedev/lightchat/core/api/ThreadLock.java
package com.github.fernthedev.lightchat.core.api;

import lombok.Getter;

public class ThreadLock {

    private final Object threadLockObject = new Object();


    @Getter
    private boolean lock = false;

    public void lock() {
        lock = true;
    }

    /**
     * Notifies all threads that the object is ready
     *
     * See {@link Object#notifyAll()}
     *
     */
    public void notifyAllThreads() {
        lock = false;
        synchronized (threadLockObject) {
            threadLockObject.notifyAll();
        }
    }

    /**
     * Waits on the {@link #threadLockObject}
     *
     * See {@link Object#wait()}
     *
     * @throws InterruptedException
     */
    public void waitOnLock() throws InterruptedException {
        threadLockObject.wait();
    }

    /**
     * Waits on the {@link #threadLockObject}
     *
     * See {@link Object#wait(long)}
     *
     * @throws InterruptedException
     */
    public void waitOnLock(long timeoutMillis) throws InterruptedException {
        threadLockObject.wait(timeoutMillis);

    }

    /**
     * Waits on the {@link #threadLockObject}
     *
     * See {@link Object#wait(long, int)}
     *
     * @throws InterruptedException
     */
    public void waitOnLock(long timeoutMillis, int nanos) throws InterruptedException {
        threadLockObject.wait(timeoutMillis, nanos);
    }


}

File: java/core/src/main/java/com/github/fernthedev/lightchat/core/api/ThreadLock.kt
package com.github.fernthedev.lightchat.core.api

class ThreadLock {
    private val threadLockObject = Any()


    var lock = false
        private set
    fun lock() {
        lock = true
    }

    /**
     * Notifies all threads that the object is ready
     *
     * See [Object.notifyAll]
     *
     */
    fun notifyAllThreads() {
        lock = false
        synchronized(threadLockObject) {
//            threadLockObject.notifyAll()
        }
    }

    /**
     * Waits on the [.threadLockObject]
     *
     * See [Object.wait]
     *
     * @throws InterruptedException
     */
    @Throws(InterruptedException::class)
    fun waitOnLock() {
        synchronized(threadLockObject) {
//            threadLockObject.wait()
        }
    }

    /**
     * Waits on the [.threadLockObject]
     *
     * See [Object.wait]
     *
     * @throws InterruptedException
     */
    @Throws(InterruptedException::class)
    fun waitOnLock(timeoutMillis: Long) {
        synchronized(threadLockObject) {
//            threadLockObject.wait(timeoutMillis)
        }
    }

    /**
     * Waits on the [.threadLockObject]
     *
     * See [Object.wait]
     *
     * @throws InterruptedException
     */
    @Throws(InterruptedException::class)
    fun waitOnLock(timeoutMillis: Long, nanos: Int) {
        synchronized(threadLockObject) {
//            threadLockObject.wait(timeoutMillis, nanos)
        }
    }
}